primitives/trie/src/cache/shared_cache.rs

```diff
@@ -0,0 +1,632 @@
+			// that the internal map of the [`LruCache`] doesn't grow too much.
+			while self.maximum_cache_size.exceeds(self.size_in_bytes) {
+				// This should always be `Some(_)`, otherwise something is wrong!
+				if let Some((key, node)) = self.lru.pop_lru() {
+					update_size_in_bytes(&mut self.size_in_bytes, &key, &node);
+				}
+			}
+		});
+	}
+
+	/// Reset the cache.
+	fn reset(&mut self) {
+		self.size_in_bytes = 0;
+		self.lru.clear();
+	}
+}
+
+/// The hash of [`ValueCacheKey`].
+#[derive(Eq, PartialEq, Clone, Copy)]
+pub struct ValueCacheKeyHash(u64);
+
+impl ValueCacheKeyHash {
+	pub fn from_hasher_and_storage_key(
+		mut hasher: impl std::hash::Hasher,
+		storage_key: &[u8],
+	) -> Self {
+		hasher.write(storage_key);
+
+		Self(hasher.finish())
+	}
+}
+
+impl std::hash::Hash for ValueCacheKeyHash {
+	fn hash<Hasher: std::hash::Hasher>(&self, state: &mut Hasher) {
+		state.write_u64(self.0);
+	}
+}
+
+impl nohash_hasher::IsEnabled for ValueCacheKeyHash {}
+
+/// A type that can only be constructed inside of this file.
+///
+/// It "requires" that the user has read the docs to prevent fuck ups.
+#[derive(Eq, PartialEq)]
+pub(super) struct IReadTheDocumentation(());
+
+/// The key type that is being used to address a [`CachedValue`].
+///
+/// This type is implemented as `enum` to improve the performance when accessing the value cache.
+/// The problem being that we need to calculate the `hash` of [`Self`] in worst case three times
+/// when trying to find a value in the value cache. First to lookup the local cache, then the shared
+/// cache and if we found it in the shared cache a third time to insert it into the list of accessed
+/// values. To work around each variant stores the `hash` to identify a unique combination of
+/// `storage_key` and `storage_root`. However, be aware that this `hash` can lead to collisions when
+/// there are two different `storage_key` and `storage_root` pairs that map to the same `hash`. This
+/// type also has the `Hash` variant. This variant should only be used for the use case of updating
+/// the lru for a key. Because when using only the `Hash` variant to getting a value from a hash map
+/// it could happen that a wrong value is returned when there is another key in the same hash map
+/// that maps to the same `hash`. The [`PartialEq`] implementation is written in a way that when one
+/// of the two compared instances is the `Hash` variant, we will only compare the hashes. This
+/// ensures that we can use the `Hash` variant to bring values up in the lru.
+#[derive(Eq)]
+pub(super) enum ValueCacheKey<'a, H> {
+	/// Variant that stores the `storage_key` by value.
+	Value {
+		/// The storage root of the trie this key belongs to.
+		storage_root: H,
+		/// The key to access the value in the storage.
+		storage_key: Arc<[u8]>,
+		/// The hash that identifying this instance of `storage_root` and `storage_key`.
+		hash: ValueCacheKeyHash,
+	},
+	/// Variant that only references the `storage_key`.
+	Ref {
```

Comment:
`HashMap::get` accepts any type `Q` as long as `K: Borrow<Q>`. Would it be possible to have a separate type `ValueCacheKeyRef` which is a borrowed version of `ValueCacheKey ` and avoid unsafe conversions?

Reply:
Yeah I wanted to do the same. However,  the problem is that the lru map doesn't support this. They support this only in nightly.